### PR TITLE
Refactor GCSE controllers and details form

### PR DESCRIPTION
**Pull request by davidgisbey**
## Context

While completing https://trello.com/c/ipPn4Y8v/2998-refactor-basecontrollers-to-startcontrollers-where-appropriate
it became clear that the GCSE controllers needed a good refactor. 

While doing this I've also broken the GcseQualificationDetailsForm into the GcseYearForm and the MathsGcseGradeForm.

## Changes proposed in this pull request

 - Dry out the controllers 
   - move all shared logic into a base controller 
   -  Implement a current qualification method and use it throughout the controllers
   - use the established pattern of passing params into the form object via a private params method
- Breaks the details form down into separate forms for grade and award_year

## Guidance to review

Per commit will be easiest.

I'm off next week so if people are happy with the approach and there are minimal changes required, please feel free to make them and merge. 

## Link to Trello card

https://trello.com/c/9yXwYX6c/3058-refactor-the-gcse-controllers
https://trello.com/c/ipPn4Y8v/2998-refactor-basecontrollers-to-startcontrollers-where-appropriate

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

**Review by TobyRet — APPROVED** [review]
Nice job. LGTM 👍

File: spec/forms/candidate_interface/restructured_work_history/work_history_break_form_spec.rb [hunk 1]
```diff
@@ -22,6 +22,7 @@
   describe 'validations' do
     it { is_expected.to validate_presence_of(:reason) }
 
+
```
Comment (raamSoftwire) [inline spec/forms/candidate_interface/restructured_work_history/work_history_break_form_spec.rb]: Do we want this extra line?

**Review by raamSoftwire — COMMENTED** [review]
I really like this refactor. Just a few small comments from me.

File: spec/forms/candidate_interface/restructured_work_history/work_history_break_form_spec.rb (hunk 1 above)
Comment (raamSoftwire) [reply spec/forms/candidate_interface/restructured_work_history/work_history_break_form_spec.rb]: Now outdated

File: app/forms/candidate_interface/gcse_year_form.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    def save(qualification)
+      return false unless valid?
+
+      qualification.update!(award_year: award_year)
+    end
+
+  private
+
+    def award_year_is_not_in_the_future
```
[the remark this answers is not in the document]
Comment (despo) [reply app/forms/candidate_interface/gcse_year_form.rb:25]: Agree with Ram. You should be able to do that by using the `future: true` parameter in line 7. The error message entry in the locale file should also not be necessary.

**Pull request merged**